README
NAME
    WWW::Hashbang::Pastebin::Client - a client library for
    WWW::Hashbang::Pastebin websites

VERSION
    version 0.002

SYNOPSIS
        use WWW::Hashbang::Pastebin::Client;
        my $client = WWW::Hashbang::Pastebin::Client->new(url => 'http://p.hashbang.ca');

        # retrieve paste content by paste ID
        print $client->get('b'), "\n";

        # create a paste from a string
        my $pasted_string_url = $client->paste(paste => rand());

        # create a paste from a file
        my $pasted_file_url = $client->paste(file => '/var/log/syslog');

        print "$pasted_string_url\n$pasted_file_url\n";

DESCRIPTION
    WWW::Hashbang::Pastebin::Client is, as you might expect, a client
    library for interfacing with WWW::Hashbang::Pastebin websites. It also
    ships with an example command-line client p.

METHODS
  new
    Creates a new client object. You must provide the URL of the
    WWW::Hashbang::Pastebin site you want to talk to:

        my $client = WWW::Hashbang::Pastebin::Client->new(url => 'http://p.hashbang.ca');

  paste
    Create a new paste on the specified website. Specify either "file" to
    read in the named file, or "paste" to provide the text directly:

        # create a paste from a string
        my $pasted_string_url = $client->paste(paste => rand());

        # create a paste from a file
        my $pasted_file_url = $client->paste(file => '/var/log/syslog');

        print "$pasted_string_url\n$pasted_file_url\n";

  put
    This is a synonym for "paste".

  get
    Get paste content from the pastebin. Pass just the ID of the paste:

        # retrieve paste content by paste ID
        print $client->get('b'), "\n";

  retrieve
    This is a synonym for "get"

AVAILABILITY
    The project homepage is
    <http://metacpan.org/release/WWW-Hashbang-Pastebin-Client/>.

    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <https://metacpan.org/module/WWW::Hashbang::Pastebin::Client/>.

SOURCE
    The development version is on github at
    <http://github.com/doherty/WWW-Hashbang-Pastebin-Client> and may be
    cloned from <git://github.com/doherty/WWW-Hashbang-Pastebin-Client.git>

BUGS AND LIMITATIONS
    You can make new bug reports, and view existing ones, through the web
    interface at
    <https://github.com/doherty/WWW-Hashbang-Pastebin-Client/issues>.

AUTHOR
    Mike Doherty <doherty@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by Mike Doherty.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl module {{$dist->name}}

{{$NEXT}}

0.003     2012-05-29
  - Accidental dependency on Plack::Loader removed [GH #1]
  - Replace LWP::UserAgent with HTTP::Tiny
  - Remove URI

0.002     2012-03-31
  - Update for latest changes in WWW::Hashbang::Pastebin - strip terminating + from
    ID, if present (instead of ., which was the previous way to get HTML output)

0.001     2012-03-16
  - First release


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "Module::Build" => "0.3601",
    "Plack::Loader" => 0,
    "Test::More" => "0.96"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "a client library for WWW::Hashbang::Pastebin websites",
  "dist_author" => [
    "Mike Doherty <doherty\@cpan.org>"
  ],
  "dist_name" => "WWW-Hashbang-Pastebin-Client",
  "dist_version" => "0.002",
  "license" => "perl",
  "module_name" => "WWW::Hashbang::Pastebin::Client",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Getopt::Long" => 0,
    "LWP::UserAgent" => 0,
    "Pod::Usage" => 0,
    "URI" => 0,
    "perl" => "v5.10.1",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [
    "bin/p"
  ]
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/WWW/Hashbang/Pastebin/Client.pm
package WWW::Hashbang::Pastebin::Client;
use strict;
use warnings;
# ABSTRACT: a client library for WWW::Hashbang::Pastebin websites
# VERSION

use HTTP::Tiny;
use Carp;

=head1 SYNOPSIS

    use WWW::Hashbang::Pastebin::Client;
    my $client = WWW::Hashbang::Pastebin::Client->new(url => 'http://p.hashbang.ca');

    # retrieve paste content by paste ID
    print $client->get('b'), "\n";

    # create a paste from a string
    my $pasted_string_url = $client->paste(paste => rand());

    # create a paste from a file
    my $pasted_file_url = $client->paste(file => '/var/log/syslog');

    print "$pasted_string_url\n$pasted_file_url\n";


=head1 DESCRIPTION

B<WWW::Hashbang::Pastebin::Client> is, as you  might expect, a client library
for interfacing with L<WWW::Hashbang::Pastebin> websites. It also ships with
an example command-line client L<p>.

=head1 METHODS

=head2 new

Creates a new client object. You must provide the URL of the
L<WWW::Hashbang::Pastebin> site you want to talk to:

    my $client = WWW::Hashbang::Pastebin::Client->new(url => 'http://p.hashbang.ca');

=cut

sub new {
    my $class = shift;
    my %args = @_;
    croak 'No pastebin URL provided' unless $args{url};
    croak 'Pastebin must be an absolute URL' unless $args{url} =~ m{^http};

    my $self = { url => $args{url} };
    $self->{ua} = HTTP::Tiny->new(
        agent => __PACKAGE__ . '/'
            . (__PACKAGE__->VERSION ? __PACKAGE__->VERSION : 'dev')
            . ' (+https://metacpan.org/module/' . __PACKAGE__ . ')'
    );

    return bless $self, $class;
}

=head2 paste

Create a new paste on the specified website. Specify either C<file> to read in
the named file, or C<paste> to provide the text directly:

    # create a paste from a string
    my $pasted_string_url = $client->paste(paste => rand());

    # create a paste from a file
    my $pasted_file_url = $client->paste(file => '/var/log/syslog');

    print "$pasted_string_url\n$pasted_file_url\n";

=cut

sub paste {
    my $self = shift;
    my %args = @_;

    if ($args{file}) {
        $args{paste} = do {
            local $/;
            open my $in, '<', $args{file}
                or die "Can't open $args{file} for reading: $!";
            <$in>
        };
    }
    croak 'No paste content given' unless $args{paste};

    my $post_response = $self->{ua}->post_form(
        $self->{url}, { p => $args{paste} }
    );

    return $post_response->{headers}->{'X-Pastebin-URL'} || $post_response->{content}
        if $post_response->{success};

    die $post_response->{status}, ' ' , $post_response->{reason}, ' ', $post_response->{content};
}

=head2 put

This is a synonym for L</paste>.

=cut

sub put {
    my $self = shift;
    $self->paste(@_);
}

=head2 get

Get paste content from the pastebin. Pass just the ID of the paste:

    # retrieve paste content by paste ID
    print $client->get('b'), "\n";

=cut

sub get {
    my $self = shift;
    my $id   = shift;
    croak 'No paste ID given' unless $id;
    
    $id =~ s{^\Q$self->{url}\E}{} if ($id =~ m/\Q$self->{url}\E/);
    $id =~ s{^/}{};
    $id =~ s{\+$}{};
    chomp $id;

    my $URI = "$self->{url}/$id";
    my $get_response = $self->{ua}->get($URI);

    return $get_response->{content} if $get_response->{success};

    die $get_response->{status}, ' ' , $get_response->{reason}, ' ', $get_response->{content};
}

=head2 retrieve

This is a synonym for L</get>

=cut

sub retrieve {
    my $self = shift;
    $self->get(@_);
}

1;


Makefile.PL

use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "a client library for WWW::Hashbang::Pastebin websites",
  "AUTHOR" => "Mike Doherty <doherty\@cpan.org>",
  "BUILD_REQUIRES" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "Module::Build" => "0.3601",
    "Plack::Loader" => 0,
    "Test::More" => "0.96"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "WWW-Hashbang-Pastebin-Client",
  "EXE_FILES" => [
    "bin/p"
  ],
  "LICENSE" => "perl",
  "NAME" => "WWW::Hashbang::Pastebin::Client",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Getopt::Long" => 0,
    "LWP::UserAgent" => 0,
    "Pod::Usage" => 0,
    "URI" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.002",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



